src/r1cs/sparse.rs

```diff
@@ -0,0 +1,228 @@
+  }
+
+  /// Multiply by a dense vector; uses rayon/gpu.
+  /// This does not check that the shape of the matrix/vector are compatible.
+  #[tracing::instrument(skip_all, name = "SparseMatrix::multiply_vec_unchecked")]
+  pub fn multiply_vec_unchecked(&self, vector: &[F]) -> Vec<F> {
+    self
+      .indptr
+      .par_windows(2)
+      .map(|ptrs| {
+        self
+          .get_row_unchecked(ptrs)
+          .map(|(val, col_idx)| *val * vector[*col_idx])
+          .sum()
+      })
+      .collect()
+  }
+
+  /// number of non-zero entries
+  pub fn len(&self) -> usize {
+    *self.indptr.last().unwrap()
+  }
+
+  /// empty matrix
+  pub fn is_empty(&self) -> bool {
+    self.len() == 0
+  }
+
+  /// returns a custom iterator
+  pub fn iter(&self) -> Iter<'_, F> {
+    let mut row = 0;
+    while self.indptr[row + 1] == 0 {
+      row += 1;
+    }
+    Iter {
+      matrix: self,
+      row,
+      i: 0,
+      nnz: *self.indptr.last().unwrap(),
+    }
+  }
+}
+
+/// Iterator for sparse matrix
+pub struct Iter<'a, F: PrimeField> {
+  matrix: &'a SparseMatrix<F>,
+  row: usize,
+  i: usize,
+  nnz: usize,
+}
+
+impl<'a, F: PrimeField> Iterator for Iter<'a, F> {
+  type Item = (usize, usize, F);
+
+  fn next(&mut self) -> Option<Self::Item> {
+    // are we at the end?
+    if self.i == self.nnz {
+      return None;
+    }
+
+    // compute current item
+    let curr_item = (
+      self.row,
+      self.matrix.indices[self.i],
+      self.matrix.data[self.i],
+    );
+
+    // advance the iterator
+    self.i += 1;
+    // edge case at the end
+    if self.i == self.nnz {
+      return Some(curr_item);
+    }
+    // if `i` has moved to next row
+    while self.i >= self.matrix.indptr[self.row + 1] {
+      self.row += 1;
+    }
+
+    Some(curr_item)
+  }
+}
+
+#[cfg(test)]
+mod tests {
+  use crate::traits::Group;
+
+  use super::SparseMatrix;
+  use pasta_curves::pallas::Point as G;
+
+  #[test]
+  fn test_matrix_creation() {
+    type Fr = <G as Group>::Scalar;
+    let matrix_data = vec![
+      (0, 1, Fr::from(2)),
+      (1, 2, Fr::from(3)),
+      (2, 0, Fr::from(4)),
+    ];
+    let sparse_matrix = SparseMatrix::<Fr>::new(&matrix_data, 3, 3);
+
+    assert_eq!(
+      sparse_matrix.data,
+      vec![Fr::from(2), Fr::from(3), Fr::from(4)]
+    );
+    assert_eq!(sparse_matrix.indices, vec![1, 2, 0]);
+    assert_eq!(sparse_matrix.indptr, vec![0, 1, 2, 3]);
+  }
+
+  #[test]
+  fn test_matrix_vector_multiplication() {
+    type Fr = <G as Group>::Scalar;
+    let matrix_data = vec![
+      (0, 1, Fr::from(2)),
+      (1, 2, Fr::from(3)),
+      (2, 0, Fr::from(4)),
+    ];
+    let sparse_matrix = SparseMatrix::<Fr>::new(&matrix_data, 3, 3);
+    let vector = vec![Fr::from(1), Fr::from(2), Fr::from(3)];
+
+    let result = sparse_matrix.multiply_vec(&vector);
+
+    // Assuming the multiplication is done row by row,
+    // Result[0] = 2.0 * 2.0 = 4.0
+    // Result[1] = 3.0 * 3.0 = 9.0
+    // Result[2] = 4.0 * 1.0 = 4.0
```

Comment:
This is nice, but I think it could be complemented with :
1. a test annotated with `#[should_panic]` testing that `multiply_vec` indeed checks the matrix-vector dimensions,
2. an example where there is more than one non-zero operand in a row's terms, to check the addition is performed correctly (at present this just tests multiplication). 
e.g. 
```rust
    let matrix_data = vec![
      (0, 1, Fr::from(2)),
      (0, 2, Fr::from(7)),
      (1, 2, Fr::from(3)),
      (2, 0, Fr::from(4)),
    ];
    let vector = vec![Fr::from(1), Fr::from(2), Fr::from(3)];
    // to be compared with vec![Fr::from(25), Fr::from(9), Fr::from(4)]
```

If you want to go the whole 9 yards, a proptest would be tip top, but not completely necessary.